src/test/java/org/opentripplanner/updater/trip/TimetableSnapshotSourceTest.java

```diff
@@ -1052,14 +1167,97 @@ public void repeatedlyAddedTripWithNewRoute() {
+      var secondPattern = assertAddedTrip(SERVICE_DATE, addedTripId, updater, false);
       var secondRoute = secondPattern.getRoute();
 
       // THEN
 
       assertSame(firstRoute, secondRoute);
       assertNotNull(transitModel.getTransitModelIndex().getRouteForId(firstRoute.getId()));
     }
+
+    static List<Arguments> addedRemovalTestCase() {
+      return List.of(
+        // TODO we might want to change the behaviour so that only the trip without pattern is
+        // persisted if the added trip is cancelled
+        Arguments.of(CANCELED, RealTimeState.CANCELED),
+        Arguments.of(DELETED, RealTimeState.DELETED)
+      );
+    }
+
+    @ParameterizedTest
+    @MethodSource("addedRemovalTestCase")
+    public void cancelingAddedTrip(
+      ScheduleRelationship scheduleRelationship,
+      RealTimeState expectedState
+    ) {
+      var builder = new TripUpdateBuilder(
+        addedTripId,
+        SERVICE_DATE,
+        ADDED,
+        transitModel.getTimeZone()
+      );
+
+      builder.addStopTime("A", 30).addStopTime("C", 40).addStopTime("E", 55);
+
+      var tripUpdate = builder.build();
+
+      var updater = defaultUpdater();
+
+      // WHEN
+      updater.applyTripUpdates(
+        TRIP_MATCHER_NOOP,
+        REQUIRED_NO_DATA,
+        fullDataset,
+        List.of(tripUpdate),
+        feedId
+      );
+
+      // THEN
+      assertAddedTrip(SERVICE_DATE, this.addedTripId, updater, true);
+
+      builder = new TripUpdateBuilder(addedTripId, SERVICE_DATE, ADDED, transitModel.getTimeZone());
+
+      var tripDescriptorBuilder = TripDescriptor.newBuilder();
+      tripDescriptorBuilder.setTripId(addedTripId);
+      tripDescriptorBuilder.setScheduleRelationship(scheduleRelationship);
+
+      tripDescriptorBuilder.setStartDate(ServiceDateUtils.asCompactString(SERVICE_DATE));
+      tripUpdate = TripUpdate.newBuilder().setTrip(tripDescriptorBuilder).build();
+
+      // WHEN
+      updater.applyTripUpdates(
+        TRIP_MATCHER_NOOP,
+        REQUIRED_NO_DATA,
+        fullDataset,
+        List.of(tripUpdate),
+        feedId
+      );
+
+      // THEN
+      // Get trip pattern of last (most recently added) outgoing edge
+      var snapshot = updater.getTimetableSnapshot(true);
+      var stopA = transitModel.getStopModel().getRegularStop(new FeedScopedId(feedId, "A"));
+      var patternsAtA = snapshot.getPatternsForStop(stopA);
+
+      assertNotNull(patternsAtA, "Added trip pattern should be found");
+      var tripPattern = patternsAtA.stream().findFirst().get();
+
+      final Timetable forToday = snapshot.resolve(tripPattern, SERVICE_DATE);
+      final Timetable schedule = snapshot.resolve(tripPattern, null);
+
+      assertNotSame(forToday, schedule);
+
+      final int forTodayAddedTripIndex = forToday.getTripIndex(addedTripId);
+      assertTrue(
+        forTodayAddedTripIndex > -1,
+        "Added trip should not be found in time table for service date"
```

Comment:
The assertion seems to be testing that the trip _is_ present. The test expects the trip to be present but with deleted/canceled status. So I believe this assertion message is backward.